### PR TITLE
implement connection gating at the top level

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For #872.

* This PR deprecates the public `Filter` and `FilterAddress` API in favour of the `ConnectionGater` API.
* Users can use either a `Filter` and by extension `FilterAddress` OR the `ConnectionGater` but NOT both.
* `Filters` is adapted to a `ConnectionGater` by means of `FilterConnectionGater`
* Adds the `ConenctionGater` to the "reflection magic" so it is available to Transports that need it.
  * While transports that need upgrading like TCP & WS will have their job done by the Upgrader, transports like QUIC (which do not need upgrading) should use this injected gater to gate connections.

### CODE REVIEW

File: options_filter.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// FiltersConnectionGater is an adapter that turns multiaddr.Filter into a
+// connmgr.ConnectionGater. It's not intended to be used directly by the user,
+// but it can.
+type FiltersConnectionGater ma.Filters
```
Comment (Stebalien) [inline options_filter.go:15]: @raulk shouldn't this live somewhere else? Or is there a better implementation I should use in go-ipfs?

Comment (Stebalien) [reply options_filter.go:15]: (at the moment, this lives in the go-libp2p root package)

Comment (raulk) [reply options_filter.go:15]: Yeah, there's no reason to make this public: https://github.com/libp2p/go-libp2p/pull/946

File: config/reflection_magic.go [hunk 2]
```diff
@@ -76,7 +78,7 @@ func callConstructor(c reflect.Value, args []reflect.Value) (interface{}, error)
 	return val, err
 }
 
-type constructor func(h host.Host, u *tptu.Upgrader) interface{}
+type constructor func(h host.Host, u *tptu.Upgrader, cg connmgr.ConnectionGater) interface{}
```
Comment (Stebalien) [inline config/reflection_magic.go:81]: @aarshkshah1992 

What was the motivation to change this function signature? It's not really an issue, but it's redundant with the `ConnGater` field of the upgrader.

Comment (raulk) [reply config/reflection_magic.go:81]: Yes, I had the same thought but given the types are internal, I thought it wouldn't harm us to rectify later. Wanna tackle that as a follow-up, @aarshkshah1992?

Comment (aarshkshah1992) [reply config/reflection_magic.go:81]: @Stebalien @raulk 

I think this was an oversight. I'll create a follow up PR to get rid of this. #951 .

Comment (Stebalien) [reply config/reflection_magic.go:81]: Thanks! Note: this isn't a super high priority.

**Pull request merged**